### DISCUSSION TITLE
iperf3 issue with speed ?

### DISCUSSION THREAD

**Discussion by usmansa2** [esnet/iperf#1686] [General]
Hi, 

I was doing some troubleshooting to check my connection speed with cloud when I faced a weird issue, i setup a computer on cloud platform, make it a server and then did the iperf test from my home connection, when i used the "-- bidir" from my computer the speed does not go beyond 3 to 4 Mb (my connection speed is 500 Mb) and when i dont use "-- bidir" and instead I use "-P 10", it immediately started giving me 100 to 200 Mb of speed. I am totally confused here, is this an operating system issue or an issue with iperf settings. Can someone please help ? thanks

**Comment by davidBar-On** [comment]
Hi, `--bidir` has two different throughput - from the client to the server and from the server to the client.  In which direction do you see the 3-4Mbps?  What is the throughput in the other direction?

To better understand your issue, run also using `-R` (instead of `--bidir`) which is the "reverse" mode - the server sends to the client.  What is the throughput in this case?  How it compares to the server to client throughput when using `--bidir`?

**Reply by usmansa2** [reply]
I am receiving 3 to 4 Mbps in both in-bound and out-bound direction, sometimes its 3.35 in and 3 out and sometimes 3.5 both in and out.

**Reply by usmansa2** [reply]
I will check -R, but what difference it will make and how it will be helpful in diagnosing the issue ?

**Comment by davidBar-On** [comment]
A test using `-R` can help to better understand what the issue may be.  In many networks, the throughput of the Upload (client to server) and Download (server to client) directions are different.  Therefore, using `-R` will help to understand the Download throughput which is used by the bi-directional test.

In addition, since the bi-directional loads the computers and the network similarly to running two copies of iperf3 - one without `-R` and `--bidir` and one with `-R`.  That load may effect performance.

**Reply by usmansa2** [reply]
hi, today we tried with -R but again it doesnt make any difference, the speed was the same

**Reply by davidBar-On** [reply]
Hi, in this case the only thing I can think of is that certain resource(s) was exhausted, in either the client or server machines.  E.g. processors, CPU, memory, network buffers.

If the `-P10` is used in the `--bidir` test, then try running two tests without `--bidir`: one with `-R` and one without it, but both with `-P20` (or even `-P30`).  This is because `--bidir -P10` is using 20 streams - 10 in each direction.  If there is a resource limit, then it may be happen also in one of these tests.

One more point to check is in which direction, uplink or downlink, you see the 3-Mbps when using the `--bidir` (or was in the total for both?).

**Reply by usmansa2** [reply]
Hi, when I used -P10 with --bidir the result is same but when i used just -P10 it gives the speed which I should get

**Reply by davidBar-On** [reply]
Hi, in this case, as I wrote above, the last test I can suggest is "running two tests without `--bidir`: one with `-R` and one without it, but both with `-P20` (or even `-P30`)".

If the throughput of these tests is also as expected, than CPU usage, memory usage, threads allocation to processors, etc. should probably be checked.  One more thing that may help the evaluation is logging and analyze the `--bidir` traffic between the client and server (using Wireshark, tcpdump, etc.) to try understanding what the problem may be - a lot of TCP retries, slow sending, etc.
